apps/agreement/contracts/Agreement.sol

```diff
@@ -0,0 +1,1350 @@
+        bool closed;                        // Whether the action has been closed for challenges
+        bytes context;                      // Link to a human-readable text providing context for the given action
+        uint256 currentChallengeId;         // Identification number of the action's currently open challenge, if any
+    }
+
+    struct Challenge {
+        uint256 actionId;                   // Identification number of the action associated to the challenge
+        address challenger;                 // Address that challenged the action
+        uint64 endDate;                     // Last date the submitter can raise a dispute against the challenge
+        bytes context;                      // Link to a human-readable text providing context for the challenge
+        uint256 settlementOffer;            // Amount of collateral tokens the challenger would accept without involving the arbitrator
+        uint256 arbitratorFeeAmount;        // Amount of arbitration fees paid by the challenger in advance
+        ERC20 arbitratorFeeToken;           // ERC20 token used for the arbitration fees paid by the challenger in advance
+        ChallengeState state;               // Current state of the action challenge
+        bool submitterFinishedEvidence;     // Whether the action submitter has finished submitting evidence for a raised dispute
+        bool challengerFinishedEvidence;    // Whether the action challenger has finished submitting evidence for a raised dispute
+        uint256 disputeId;                  // Identification number of the dispute on the arbitrator
+        uint256 ruling;                     // Ruling given for the action's dispute
+    }
+
+    struct CollateralRequirement {
+        ERC20 token;                        // ERC20 token to be used for collateral
+        uint256 actionAmount;               // Amount of collateral token that will be locked from the submitter's staking pool every time an action is created
+        uint256 challengeAmount;            // Amount of collateral token that will be locked from the challenger's own balance every time an action is challenged
+        uint64 challengeDuration;           // Challenge duration in seconds, during which the submitter can raise a dispute
+        Staking staking;                    // Staking pool cache for the collateral token
+    }
+
+    struct DisputableInfo {
+        bool registered;                                                    // Whether a Disputable app is registered
+        uint256 collateralRequirementsLength;                               // Identification number of the next collateral requirement instance
+        mapping (uint256 => CollateralRequirement) collateralRequirements;  // List of collateral requirements indexed by id
+    }
+
+    StakingFactory public stakingFactory;                           // Staking factory to be used for the collateral staking pools
+
+    uint256 private settingsLength;
```

Comment:
Maybe better to name these as `nextSettingsId`/`currentSettingsId` instead, since we are using them as IDs elsewhere.

I think it might also make the code a tad easier to work with if we go with `current`, as we'll avoid having to do a `-1` every time we need it.

Reply:
I tried using current before, but it doesn't fit properly for the first case where it is initialized as 0 and there is actually no instance stored yet. Let's go with `next*Id` for now :)